File: js/packages/botbuilder-m365/src/AI.ts
/**
 * @module botbuilder-m365
 */
/**
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License.
 */

import { CardFactory, Channels, MessageFactory, TurnContext } from 'botbuilder';
import { ConversationHistory } from './ConversationHistory';
import { DefaultModerator } from './DefaultModerator';
import { DefaultTempState, DefaultTurnState } from './DefaultTurnStateManager';
import { Moderator } from './Moderator';
import { PredictedDoCommand, PredictedSayCommand, Planner, Plan } from './Planner';
import { PromptManager, PromptTemplate } from './Prompts';
import { ResponseParser } from './ResponseParser';
import { TurnState } from './TurnState';

export type PromptSelector<TState extends TurnState> = (Context: TurnContext, state: TState) => Promise<string|PromptTemplate>;

export interface PredictedDoCommandAndHandler<TState> extends PredictedDoCommand {
    handler: (context: TurnContext, state: TState, data?: Record<string, any>, action?: string) => Promise<boolean>;
}

export interface AIOptions<TState extends TurnState> {
    planner: Planner<TState>;
    promptManager: PromptManager<TState>;
    moderator?: Moderator<TState>;
    prompt?: string|PromptTemplate|PromptSelector<TState>;
    history?: Partial<AIHistoryOptions>;
}

export interface AIHistoryOptions {
    trackHistory: boolean;
    maxTurns: number;
    maxTokens: number;
    lineSeparator: string;
    userPrefix: string;
    assistantPrefix: string;
    assistantHistoryType: 'text' | 'planObject';
}

export interface ConfiguredAIOptions<TState extends TurnState> {
    planner: Planner<TState>;
    promptManager: PromptManager<TState>;
    moderator: Moderator<TState>;
    prompt?: string|PromptTemplate|((Context: TurnContext, state: TState) => Promise<string|PromptTemplate>);
    history: AIHistoryOptions;
}

export class AI<TState extends TurnState = DefaultTurnState> {
    private readonly _actions: Map<string, ActionEntry<TState>> = new Map();
    private readonly _options: ConfiguredAIOptions<TState>;

    public static readonly UnknownActionName = '___UnknownAction___';
    public static readonly FlaggedInputActionName = '___FlaggedInput___';
    public static readonly FlaggedOutputActionName = '___FlaggedOutput___';
    public static readonly RateLimitedActionName = '___RateLimited___';
    public static readonly PlanReadyActionName = '___PlanReady___';
    public static readonly DoCommandActionName = '___DO___';
    public static readonly SayCommandActionName = '___SAY___';

    public constructor(options: AIOptions<TState>) {
        this._options = Object.assign({}, options) as ConfiguredAIOptions<TState>;

        // Create moderator if needed
        if (!this._options.moderator) {
            this._options.moderator = new DefaultModerator<TState>();
        }

        // Initialize history options
        this._options.history = Object.assign({
            trackHistory: true,
            maxTurns: 3,
            maxTokens: 1000,
            lineSeparator: '\n',
            userPrefix: 'User:',
            assistantPrefix: 'Assistant:',
            assistantHistoryType: 'planObject'
        } as AIHistoryOptions, this._options.history);

        // Register default UnknownAction handler
        this.action(
            AI.UnknownActionName,
            (context, state, data, action) => {
                console.error(`An AI action named "${action}" was predicted but no handler was registered.`);
                return Promise.resolve(true);
            },
            true
        );

        // Register default FlaggedInputAction handler
        this.action(
            AI.FlaggedInputActionName,
            (context, state, data, action) => {
                console.error(
                    `The users input has been moderated but no handler was registered for 'AI.FlaggedInputActionName'.`
                );
                return Promise.resolve(true);
            },
            true
        );

        // Register default FlaggedOutputAction handler
        this.action(
            AI.FlaggedOutputActionName,
            (context, state, data, action) => {
                console.error(
                    `The bots output has been moderated but no handler was registered for 'AI.FlaggedOutputActionName'.`
                );
                return Promise.resolve(true);
            },
            true
        );

        // Register default RateLimitedActionName
        this.action(
            AI.RateLimitedActionName,
            (context, state, data, action) => {
                throw new Error(`An AI request failed because it was rate limited`);
            },
            true
        );

        // Register default PlanReadyActionName
        this.action<Plan>(
            AI.PlanReadyActionName,
            async (context, state, plan) => {
                return Array.isArray(plan.commands) && plan.commands.length > 0;
            },
            true
        );

        // Register default DoCommandActionName
        this.action<PredictedDoCommandAndHandler<TState>>(
            AI.DoCommandActionName,
            async (context, state, data, action) => {
                const { entities, handler } = data;
                return await handler(context, state, entities, action);
            },
            true
        );

        // Register default SayCommandActionName
        this.action<PredictedSayCommand>(
            AI.SayCommandActionName,
            async (context, state, data, action) => {
                const response = data.response;
                const card = ResponseParser.parseAdaptiveCard(response);
                if (card) {
                    const attachment = CardFactory.adaptiveCard(card);
                    const activity = MessageFactory.attachment(attachment);
                    await context.sendActivity(activity);
                } else if (context.activity.channelId == Channels.Msteams) {
                    await context.sendActivity(response.split('\n').join('<br>'));
                } else {
                    await context.sendActivity(response);
                }

                return true;
            },
            true
        );
    }

    public get moderator(): Moderator<TState> {
        return this._options.moderator;
    }

    public get options(): ConfiguredAIOptions<TState> {
        return this._options;
    }

    public get planner(): Planner<TState> {
        return this._options.planner;
    }

    public get prompts(): PromptManager<TState> {
        return this._options.promptManager;
    }

    /**
     * Registers a handler for a named action.
     *
     *
     * Actions can be triggered by a planner returning a DO command.
     *
     * @param name Unique name of the action.
     * @param handler Function to call when the action is triggered.
     * @param allowOverrides Optional. If true
     * @returns The application instance for chaining purposes.
     */
    public action<TEntities = Record<string, any>>(
        name: string | string[],
        handler: (context: TurnContext, state: TState, entities: TEntities, action: string) => Promise<boolean>,
        allowOverrides = false
    ): this {
        (Array.isArray(name) ? name : [name]).forEach(n => {
            if (!this._actions.has(n) || allowOverrides) {
                this._actions.set(n, { handler, allowOverrides });
            } else {
                const entry = this._actions.get(n);
                if (entry!.allowOverrides) {
                    entry!.handler = handler;
                } else {
                    throw new Error(
                        `The AI.action() method was called with a previously registered action named "${n}".`
                    );
                }
            }
        });

        return this;
    }

    public async chain(
        context: TurnContext,
        state: TState,
        prompt?: string|PromptTemplate,
        options?: Partial<AIOptions<TState>>
    ): Promise<boolean> {
        // Configure options
        const opts = this.configureOptions(options);

        // Select prompt
        if (!prompt) {
            if (opts.prompt == undefined) {
                throw new Error(`AI.chain() was called without a prompt and no default prompt was configured.`);
            } else if (typeof opts.prompt == 'function') {
                prompt = await opts.prompt(context, state);
            } else {
                prompt = opts.prompt;
            }
        }

        // Populate {{$temp.input}}
        const temp = (state as any as DefaultTurnState)?.temp?.value ?? {} as DefaultTempState;
        if (typeof temp.input != 'string') {
            // Use the received activity text
            temp.input = context.activity.text;
        }

        // Populate {{$temp.history}}
        if (typeof temp.history != 'string' && opts.history.trackHistory) {
            temp.history = ConversationHistory.toString(state, opts.history.maxTokens, opts.history.lineSeparator);
        }

        // Render the prompt
        const renderedPrompt = await opts.promptManager.renderPrompt(context, state, prompt);

        // Generate plan
        let plan = await opts.moderator.reviewPrompt(context, state, renderedPrompt, opts);
        if (!plan) {
            plan = await opts.planner.generatePlan(context, state, renderedPrompt, opts);
            plan = await opts.moderator.reviewPlan(context, state, plan);
        }

        // Process generated plan
        let continueChain = await this._actions
            .get(AI.PlanReadyActionName)!
            .handler(context, state, plan, '');
        if (continueChain) {
            // Update conversation history
            if (opts.history.trackHistory) {
                ConversationHistory.addLine(state, `${opts.history.userPrefix.trim()} ${temp.input.trim()}`, opts.history.maxTurns * 2);
                switch (opts.history.assistantHistoryType) {
                    case 'text':
                        // Extract only the things the assistant has said
                        const text = plan.commands.filter(v => v.type == 'SAY').map(v => (v as PredictedSayCommand).response).join('\n');
                        ConversationHistory.addLine(state, `${opts.history.assistantPrefix.trim()} ${text}`, opts.history.maxTurns * 2);
                        break;
                    case 'planObject':
                    default:
                        // Embed the plan object to re-enforce the model
                        // - TODO: add support for XML as well
                        ConversationHistory.addLine(state, `${opts.history.assistantPrefix.trim()} ${JSON.stringify(plan)}`, opts.history.maxTurns * 2);
                        break;
                }
            }

            // Run predicted commands
            for (let i = 0; i < plan.commands.length && continueChain; i++) {
                const cmd = plan.commands[i];
                switch (cmd.type) {
                    case 'DO':
                        const { action } = cmd as PredictedDoCommand;
                        if (this._actions.has(action)) {
                            // Call action handler
                            const handler = this._actions.get(action)!.handler;
                            continueChain = await this._actions
                                .get(AI.DoCommandActionName)!
                                .handler(context, state, { handler, ...(cmd as PredictedDoCommand) }, action);
                        } else {
                            // Redirect to UnknownAction handler
                            continueChain = await this._actions
                                .get(AI.UnknownActionName)!
                                .handler(context, state, plan, action);
                        }
                        break;
                    case 'SAY':
                        continueChain = await this._actions
                            .get(AI.SayCommandActionName)!
                            .handler(context, state, cmd, AI.SayCommandActionName);
                        break;
                    default:
                        throw new Error(`Application.run(): unknown command of '${cmd.type}' predicted.`);
                }
            }
        }

        return continueChain;
    }

    public async completePrompt(
        context: TurnContext,
        state: TState,
        prompt: string|PromptTemplate,
        options?: Partial<AIOptions<TState>>
    ): Promise<string|undefined> {
        // Configure options
        const opts = this.configureOptions(options);

        // Render the prompt
        const renderedPrompt = await opts.promptManager.renderPrompt(context, state, prompt);

        // Complete the prompt
        return await opts.planner.completePrompt(context, state, renderedPrompt, opts);
    }

    public createSemanticFunction(name: string, template?: PromptTemplate, options?: Partial<AIOptions<TState>>): (context: TurnContext, state: TState) => Promise<any> {
        // Cache prompt template if being dynamically assigned
        if (template) {
            this._options.promptManager.addPromptTemplate(name, template);
        }

        return (context: TurnContext, state: TState) => this.completePrompt(context, state, name, options);
    }

    public doAction<TData = Record<string, any>>(
        context: TurnContext,
        state: TState,
        action: string,
        data?: TData
    ): Promise<boolean> {
        if (!this._actions.has(action)) {
            throw new Error(`Can't find an action named '${action}'.`);
        }

        const handler = this._actions.get(action)!.handler;
        return handler(context, state, data, action);
    }

    private configureOptions(options?: Partial<AIOptions<TState>>): ConfiguredAIOptions<TState> {
        let configuredOptions: ConfiguredAIOptions<TState>;
        if (options) {
            configuredOptions = Object.assign({}, this._options,  options) as ConfiguredAIOptions<TState>;
            if (options.history) {
                // Just inherit any missing history settings
                options.history = Object.assign({}, this._options.history, options.history);
            } else {
                // Disable history tracking by default
                options.history = Object.assign({}, this._options.history, { trackHistory: false });
            }
        } else {
            configuredOptions = this._options;
        }

        return configuredOptions;
    }
}

interface ActionEntry<TState> {
    handler: (context: TurnContext, state: TState, data?: Record<string, any>, action?: string) => Promise<boolean>;
    allowOverrides: boolean;
}

File: dotnet/packages/Microsoft.Bot.Builder.M365/Microsoft.Bot.Builder.M365/AI/AI.cs
using Microsoft.Bot.Builder.M365.AI.Action;
using Microsoft.Bot.Builder.M365.AI.Planner;
using System.Reflection;
using Microsoft.Bot.Builder.M365.AI.Prompt;
using Microsoft.Bot.Builder.M365.Exceptions;
using Microsoft.Extensions.Logging;
using Microsoft.Bot.Builder.M365.Utilities;

namespace Microsoft.Bot.Builder.M365.AI
{
    /// <summary>
    /// AI System.
    /// </summary>
    /// <remarks>
    /// The AI system is responsible for generating plans, moderating input and output, and
    /// generating prompts. It can be used free standing or routed to by the Application object.
    /// </remarks>
    /// <typeparam name="TState">Optional. Type of the turn state.</typeparam>
    public class AI<TState> where TState : TurnState
    {
        private readonly IActionCollection<TState> _actions;
        private readonly AIOptions<TState> _options;

        public AI(AIOptions<TState> options, ILogger? logger = null)
        {
            Verify.NotNull(options, nameof(options));

            _options = options;
            _actions = new ActionCollection<TState>();

            if (_options.Moderator == null)
            {
                // TODO: Set default moderator
            }

            _options.History ??= new AIHistoryOptions();


            // Import default actions
            ImportActions(new DefaultActions<TState>(logger));
        }

        /// <summary>
        /// Returns the moderator being used by the AI system.
        /// </summary>
        /// <remarks>
        /// The default moderator simply allows all messages and plans through without intercepting them.
        /// </remarks>
        public Moderator<TState> Moderator => _options.Moderator;

        /// <summary>
        /// Returns the options for the AI system.
        /// </summary>
        public AIOptions<TState> Options => _options;

        /// <summary>
        /// Returns the planner being used by the AI system.
        /// </summary>
        public IPlanner<TState> Planner => _options.Planner;

        public IPromptManager<TState> Prompts => _options.PromptManager;

        /// <summary>
        /// Registers a handler for a named action.
        /// </summary>
        /// <remarks>
        /// The AI systems planner returns plans that are made up of a series of commands or actions
        /// that should be performed. Registering a handler lets you provide code that should be run in
        /// response to one of the predicted actions.
        /// 
        /// Plans support a DO command which specifies the name of an action to call and an optional
        /// set of entities that should be passed to the action. The internal plan executor will call
        /// the registered handler for the action passing in the current context, state, and entities.
        /// 
        /// Additionally, the AI system itself uses actions to handle things like unknown actions,
        /// flagged input, and flagged output. You can override these actions by registering your own
        /// handler for them. The names of the built-in actions are available as static properties on
        /// the AI class.
        /// </remarks>
        /// <param name="name">The name of the action.</param>
        /// <param name="handler">The action handler function.</param>
        /// <param name="allowOverrides">Whether or not this action's properties can be overriden.</param>
        /// <returns>The current instance object.</returns>
        /// <exception cref="Exception"></exception>
        public AI<TState> RegisterAction(string name, ActionHandler<TState> handler, bool allowOverrides = false)
        {
            Verify.NotNull(name, nameof(name));
            Verify.NotNull(handler, nameof(handler));

            if (!_actions.HasAction(name) || allowOverrides)
            {
                _actions.SetAction(name, handler, allowOverrides);
            } else
            {
                ActionEntry<TState> entry = _actions.GetAction(name);
                if (entry.AllowOverrides)
                {
                    entry.Handler = handler;
                } else
                {
                    throw new Exception($"Attempting to register an already existing action `{name}` that does not allow overrides.");
                }
            }

            return this;
        }

        /// <summary>
        /// Register an action into the AI module.
        /// </summary>
        /// <param name="action"></param>
        /// <returns>The current instance object.</returns>
        public AI<TState> RegisterAction(ActionEntry<TState> action)
        {
            Verify.NotNull(action, nameof(action));

            return RegisterAction(action.Name, action.Handler, action.AllowOverrides);
        }

        /// <summary>
        /// Import a set of Actions from the given class instance. The functions must have the `Action` attribute.
        /// Once these functions are imported, the AI module will have access to these functions.
        /// </summary>
        /// <param name="instance">Instance of a class containing these functions.</param>
        /// <returns>The current instance object.</returns>
        public AI<TState> ImportActions(object instance)
        {
            Verify.NotNull(instance, nameof(instance));

            MethodInfo[] methods = instance.GetType()
                .GetMethods(BindingFlags.Static | BindingFlags.Instance | BindingFlags.Public | BindingFlags.InvokeMethod);

            // Filter out null functions
            IEnumerable<ActionEntry<TState>> functions = from method in methods select ActionEntry<TState>.FromNativeMethod(method, instance);
            List<ActionEntry<TState>> result = (from function in functions where function != null select function).ToList();

            // Fail if two functions have the same name
            var uniquenessCheck = new HashSet<string>(from x in result select x.Name, StringComparer.OrdinalIgnoreCase);
            if (result.Count > uniquenessCheck.Count)
            {
                throw new Exception("Function overloads are not supported, please differentiate function names");
            }

            // Register the actions
            result.ForEach(action => RegisterAction(action));

            return this;
        }

        /// <summary>
        /// Chains into another prompt and executes the plan that is returned.
        /// </summary>
        /// <remarks>
        /// This method is used to chain into another prompt. It will call the prompt manager to
        /// get the plan for the prompt and then execute the plan. The return value indicates whether
        /// that plan was completely executed or not, and can be used to make decisions about whether the
        /// outer plan should continue executing.
        /// </remarks>
        /// <param name="turnContext">Current turn context.</param>
        /// <param name="turnState">Current turn state.</param>
        /// <param name="prompt">Optional. Prompt name or prompt template to use. If omitted, the AI systems default prompt will be used.</param>
        /// <param name="options">Optional. Override options for the prompt. If omitted, the AI systems configured options will be used.</param>
        /// <param name="cancellationToken">A cancellation token that can be used by other objects
        /// or threads to receive notice of cancellation.</param>
        /// <returns>True if the plan was completely executed, otherwise false.</returns>
        /// <exception cref="AIException">This exception is thrown when an unknown (not  DO or SAY) command is predicted.</exception>
        public async Task<bool> ChainAsync(ITurnContext turnContext, TState turnState, string? prompt = null, AIOptions<TState>? options = null, CancellationToken cancellationToken = default)
        {
            Verify.NotNull(turnContext, nameof(turnContext));
            Verify.NotNull(turnState, nameof(turnState));

            AIOptions<TState> aIOptions = _ConfigureOptions(options);

            // Select prompt
            if (prompt == null)
            {
                if (aIOptions.Prompt == null)
                {
                    throw new AIException("AI.ChainAsync() was called without a prompt and no default prompt was configured.");
                } else
                {
                    prompt = aIOptions.Prompt;
                }
            }

            // TODO: Populate {{$temp.input}}

            // TODO: Populate {{$temp.history}}

            // Render the prompt
            PromptTemplate renderedPrompt = await aIOptions.PromptManager.RenderPrompt(turnContext, turnState, prompt);

            return await ChainAsync(turnContext, turnState, renderedPrompt, options, cancellationToken);
        }

        /// <summary>
        /// Chains into another prompt and executes the plan that is returned.
        /// </summary>
        /// <remarks>
        /// This method is used to chain into another prompt. It will call the prompt manager to
        /// get the plan for the prompt and then execute the plan. The return value indicates whether
        /// that plan was completely executed or not, and can be used to make decisions about whether the
        /// outer plan should continue executing.
        /// </remarks>
        /// <param name="turnContext">Current turn context.</param>
        /// <param name="turnState">Current turn state.</param>
        /// <param name="prompt">Optional. Prompt template to use.</param>
        /// <param name="options">Optional. Override options for the prompt. If omitted, the AI systems configured options will be used.</param>
        /// <param name="cancellationToken">A cancellation token that can be used by other objects
        /// or threads to receive notice of cancellation.</param>
        /// <returns>True if the plan was completely executed, otherwise false.</returns>
        /// <exception cref="AIException">This exception is thrown when an unknown (not  DO or SAY) command is predicted.</exception>
        public async Task<bool> ChainAsync(ITurnContext turnContext, TState turnState, PromptTemplate prompt, AIOptions<TState>? options = null, CancellationToken cancellationToken = default)
        {
            Verify.NotNull(turnContext, nameof(turnContext));
            Verify.NotNull(turnState, nameof(turnState));
            Verify.NotNull(prompt, nameof(prompt));

            AIOptions<TState> aIOptions = _ConfigureOptions(options);

            // TODO: Populate {{$temp.input}}

            // TODO: Populate {{$temp.history}}

            // Render the prompt
            PromptTemplate renderedPrompt = await aIOptions.PromptManager.RenderPrompt(turnContext, turnState, prompt);

            // TODO: Implement moderator

            // Generate plan
            Plan plan = await aIOptions.Planner.GeneratePlanAsync(turnContext, turnState, renderedPrompt, aIOptions, cancellationToken);

            // Process generated plan
            bool continueChain = await _actions.GetAction(DefaultActionTypes.PlanReadyActionName)!.Handler(turnContext, turnState, plan);
            if (continueChain)
            {
                // TODO: Update conversation history
            }

            for (int i = 0; i < plan.Commands.Count && continueChain; i++)
            {
                IPredictedCommand command = plan.Commands[i];

                if (command is PredictedDoCommand doCommand)
                {
                    if (_actions.HasAction(doCommand.Action))
                    {
                        // Call action handler
                        continueChain = await _actions
                            .GetAction(DefaultActionTypes.DoCommandActionName)!
                            .Handler(turnContext, turnState, doCommand, doCommand.Action);
                    }
                    else
                    {
                        // Redirect to UnknownAction handler
                        continueChain = await _actions
                            .GetAction(DefaultActionTypes.UnknownActionName)
                            .Handler(turnContext, turnState, plan, doCommand.Action);
                    }
                }
                else if (command is PredictedSayCommand sayCommand)
                {
                    continueChain = await _actions
                        .GetAction(DefaultActionTypes.SayCommandActionName)
                        .Handler(turnContext, turnState, sayCommand, DefaultActionTypes.SayCommandActionName);
                }
                else
                {
                    throw new AIException($"Unknown command of {command.Type} predicted");
                }
            }

            return continueChain;
        }

        /// <summary>
        /// A helper method to complete a prompt using the configured prompt manager.
        /// </summary>
        /// <param name="turnContext">Current turn context.</param>
        /// <param name="turnState">Current turn state.</param>
        /// <param name="promptTemplate">Prompt template to use.</param>
        /// <param name="options">Optional. Override options for the prompt. If omitted, the AI systems configured options will be used.</param>
        /// <param name="cancellationToken">A cancellation token that can be used by other objects
        /// or threads to receive notice of cancellation.</param>
        public async Task<string> CompletePromptAsync(ITurnContext turnContext, TState turnState, PromptTemplate promptTemplate, AIOptions<TState>? options, CancellationToken cancellationToken)
        {
            Verify.NotNull(turnContext, nameof(turnContext));
            Verify.NotNull(turnState, nameof(turnState));
            Verify.NotNull(promptTemplate, nameof(promptTemplate));

            // Configure options
            AIOptions<TState> aiOptions = _ConfigureOptions(options);

            // Render the prompt
            PromptTemplate renderedPrompt = await aiOptions.PromptManager.RenderPrompt(turnContext, turnState, promptTemplate);

            // Complete the prompt
            return await aiOptions.Planner.CompletePromptAsync(turnContext, turnState, renderedPrompt, aiOptions, cancellationToken);
        }

        /// <summary>
        /// A helper method to complete a prompt using the configured prompt manager.
        /// </summary>
        /// <param name="turnContext">Current turn context.</param>
        /// <param name="turnState">Current turn state.</param>
        /// <param name="name">Prompt name to use</param>
        /// <param name="options">Optional. Override options for the prompt. If omitted, the AI systems configured options will be used.</param>
        /// <param name="cancellationToken">A cancellation token that can be used by other objects
        /// or threads to receive notice of cancellation.</param>
        public async Task<string> CompletePromptAsync(ITurnContext turnContext, TState turnState, string name, AIOptions<TState>? options, CancellationToken cancellationToken)
        {
            Verify.NotNull(turnContext, nameof(turnContext));
            Verify.NotNull(turnState, nameof(turnState));
            Verify.NotNull(name, nameof(name));

            // Configure options
            AIOptions<TState> aiOptions = _ConfigureOptions(options);

            // Render the prompt
            PromptTemplate renderedPrompt = await aiOptions.PromptManager.RenderPrompt(turnContext, turnState, name);

            // Complete the prompt
            return await aiOptions.Planner.CompletePromptAsync(turnContext, turnState, renderedPrompt, aiOptions, cancellationToken);
        }

        /// <summary>
        /// Creates a semantic function that can be registered with the app's prompt manager.
        /// </summary>
        /// <remarks>
        /// Semantic functions are functions that make model calls and return their results as template
        /// parameters to other prompts. For example, you could define a semantic function called
        /// 'translator' that first translates the user's input to English before calling your main prompt:
        /// <br/><br/>
        /// <c>
        /// app.AI.Prompts.AddFunction('translator', app.AI.CreateSemanticFunction('translator-prompt'));
        /// </c>
        /// <br/><br/>
        /// You would then create a prompt called "translator-prompt" that does the translation and then in
        /// your main prompt you can call it using the template expression `{{translator}}`.
        /// </remarks>
        /// <param name="name">Prompt to use. If template is provided then this name will be assigned to it in the prompt manager.</param>
        /// <param name="template">Optional. Prompt template to use.</param>
        /// <param name="options">Optional. Override options for the prompt. If omitted, the AI systems configured options will be used.</param>
        /// <returns>A prompt function.</returns>
        public PromptFunction<TState> CreateSemanticFunction(string name, PromptTemplate? template, AIOptions<TState>? options)
        {
            Verify.NotNull(name, nameof(name));

            if (template != null)
            {
                _options.PromptManager.AddPromptTemplate(name, template);
            }

            return (ITurnContext turnContext, TState turnState) => CompletePromptAsync(turnContext, turnState, name, options, default);
        }

        private AIOptions<TState> _ConfigureOptions(AIOptions<TState>? options)
        {
            AIOptions<TState> configuredOptions;
            
            if (options != null)
            {
                configuredOptions = options;

                // Disable history tracking by default
                options.History ??= new AIHistoryOptions() { TrackHistory = false };
            } else
            {
                configuredOptions = _options;
            }

            return configuredOptions;
        }
    }   
}
